src/ecmult_multi_impl.h

```diff
@@ -0,0 +1,202 @@
+    size_t root = heap->size >> 1;;
+    while (root-- > 0) {
+        secp256k1_sift_down(heap, root, heap->tree[root]);
+    }
+}
+
+static void secp256k1_heap_initialize(secp256k1_scalar_heap *heap, const secp256k1_scalar *scalars, const secp256k1_gej *pt, size_t n) {
+    size_t i, size = 0;
+
+    VERIFY_CHECK(n <= SECP256K1_ECMULT_MULTI_MAX_N);
+
+    for (i = 0; i < n; ++i) {
+        if (!secp256k1_scalar_is_zero(&scalars[i]) && !secp256k1_gej_is_infinity(&pt[i])) {
+            heap->tree[size++] = i;
+        }
+    }
+
+    heap->scalars = scalars;
+    heap->size = size;
+
+    secp256k1_heapify(heap);
+}
+
+SECP256K1_INLINE static unsigned char secp256k1_replace(secp256k1_scalar_heap *heap, unsigned char index) {
+    unsigned char result = heap->tree[0];
+    VERIFY_CHECK(heap->size > 0);
+    secp256k1_sift_floyd(heap, 0, index);
+    return result;
+}
+
+SECP256K1_INLINE static unsigned char secp256k1_heap_remove(secp256k1_scalar_heap *heap) {
+    unsigned char result = heap->tree[0];
+    VERIFY_CHECK(heap->size > 0);
+    if (--heap->size > 0) {
+        secp256k1_sift_down(heap, 0, heap->tree[heap->size]);
+    }
+    return result;
+}
+
+/** Multi-multiply: R = sum_i ni * Ai */
+static void secp256k1_ecmult_multi(secp256k1_gej *r, secp256k1_scalar *sc, secp256k1_gej *pt, size_t n) {
+    secp256k1_scalar_heap heap;
+    unsigned char first, second;
+
+    secp256k1_gej_set_infinity(r);
+    secp256k1_heap_initialize(&heap, sc, pt, n);
+
+    if (heap.size == 0) {
+        return;
+    }
+
+    first = secp256k1_heap_remove(&heap);
+
+    while (heap.size > 0) {
+        second = heap.tree[0];        
+
+        do {
+            /* Observe that nX + mY = (n-m)X + m(X + Y), and if n > m this transformation
+             * reduces the magnitude of the larger scalar, on average by half. So by
+             * repeating this we will quickly zero out all but one exponent, which will
+             * be small. */
+            secp256k1_gej_add_var(&pt[second], &pt[first], &pt[second], NULL);  /* Y -> X + Y */
+            secp256k1_scalar_numsub(&sc[first], &sc[first], &sc[second]);  /* n -> n - m */
+
+            if (secp256k1_scalar_cmp_var(&sc[first], &sc[second]) < 0) {
+                break;
+            }
+
+            /* To handle pathological inputs, we use a binary ladder step here */
+            if (secp256k1_scalar_shr_int(&sc[first], 1) == 1) {
+                secp256k1_gej_add_var(r, r, &pt[first], NULL);
+            }
+            secp256k1_gej_double_var(&pt[first], &pt[first], NULL);
+        }
+        while (secp256k1_scalar_cmp_var(&sc[first], &sc[second]) >= 0);
+
+        if (secp256k1_scalar_is_zero(&sc[first])) {
+            first = secp256k1_heap_remove(&heap);
+        } else {
+            first = secp256k1_replace(&heap, first);
+        }
+    }
+
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&sc[first]));
+
+    if (secp256k1_gej_is_infinity(&pt[first])) {
```

Comment:
It seems to me this case is a) unlikely, b) safe to handle below, c) fast if handled below anyway. So I'd suggest removing this special handling.